src/network_service/sending_targets_cache.rs

```diff
@@ -53,102 +69,129 @@ impl SendingTargetsCache {
-                TargetState::Failed(x) => Some((info, *x, state)),
-                _ => None,
-            })
-            .min_by_key(|(_info, num, _state)| *num)
-            .map(|(info, num, state)| {
-                *state = TargetState::Sending(num);
-                info
-            })
-            .cloned()
-    }
+    /// Mark the given target as succeeded in sending the given message.
+    pub fn target_succeeded(&mut self, token: Token, target_addr: SocketAddr) {
+        let mut entry = if let Entry::Occupied(entry) = self.cache.entry(token) {
+            entry
+        } else {
+            return;
+        };
 
-    fn should_drop(&self, token: Token) -> bool {
-        // Other methods maintain the invariant that exactly one of these is true:
-        // - some target is in the Sending state
-        // - we succeeded (no further sending needed)
-        // - we failed (no more targets available)
-        // So if none are sending, the handling of the message is finished and we can drop it
-        self.target_states(token)
-            .all(|(_info, state)| !state.is_sending())
-    }
+        entry
+            .get_mut()
+            .retain(|(info, _)| info.peer_addr != target_addr);
 
-    pub fn target_failed(&mut self, token: Token, target: SocketAddr) -> Option<ConnectionInfo> {
-        self.fail_target(token, target);
-        let result = self.take_next_target(token);
-        if self.should_drop(token) {
-            let _ = self.cache.remove(&token);
+        if entry
+            .get()
+            .iter()
+            .all(|(_, target)| !target.state.is_sending())
+        {
+            let _ = entry.remove();
         }
-        result
     }
+}
+
+fn set_target_failed(
+    targets: &mut Vec<(ConnectionInfo, Target)>,
+    failed_target_addr: &SocketAddr,
+) -> bool {
+    let (index, target) = if let Some((index, (_, target))) = targets
+        .iter_mut()
+        .enumerate()
+        .find(|(_, (info, _))| info.peer_addr == *failed_target_addr)
+    {
+        (index, target)
+    } else {
+        log_or_panic!(
+            LogLevel::Error,
+            "Got a failure from unknown target: {}",
+            failed_target_addr
+        );
+        return false;
+    };
 
-    pub fn target_succeeded(&mut self, token: Token, target: SocketAddr) {
-        let _ = self
-            .target_states_mut(token)
-            .find(|(info, _state)| info.peer_addr == target)
-            .map(|(_info, state)| {
-                *state = TargetState::Sent;
-            });
-        if self.should_drop(token) {
-            let _ = self.cache.remove(&token);
+    match target.state {
+        TargetState::Failed { .. } => {
+            log_or_panic!(
+                LogLevel::Error,
+                "Got a failure from failed target: {}",
+                failed_target_addr
+            );
+            true
+        }
+        TargetState::Sending(start) => {
+            target.total += start.elapsed();
```

Comment:
So this `start` is the time when this target being seclected for a sending attempt.
which means the `total` actually is the : accumulated sum of the failure detection period, i.e.  `start.elapsed()` means the time that elapsed since we send message to the target to the time we received the failure from it. 

This makes `RESEND_TIMEOUT` doesn't mean a fixed interval (say a fixed 30s since first sending attempt), but a network performance dependent factor? i.e. in a fast network there could be thousands of attempts been made?

Or did I misunderstand anything?

Reply:
No, you are right. Good point. Wonder how much of an issue this is...